File: python/exo5_3.py
import threading
import random

class TypeTableau :
    def __init__(self) :
        self.tab = []
        self.nbElements = 0
        self.x = 0

def creerTableau(a) :
        a.tab = [random.randint(0, 99) for i in range(a.nbElements)]

def chercheX(a) :
        for i in range(a.nbElements) :
            if a.tab[i] == a.x :
                resultat[0] = 1
                return
        resultat[0] = 0

A = TypeTableau()
resultat = [None]

A.nbElements = random.randint(1, 10)
th1 = threading.Thread(target=creerTableau, args=(A,))
th1.start()
A.x = int(input("Nombre à chercher: "))
th1.join()

for i in range(A.nbElements) :
    print(A.tab[i], end="  ")
print()

th2 = threading.Thread(target=chercheX, args=(A,))
th2.start()
th2.join()

if resultat[0] :
    print(f"{A.x} est trouvé.")
else :
    print(f"{A.x} n'est pas trouvé.")

File: C/exo5_3.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

typedef struct
{
    int *tab;
    int nbElements;
    int x;
}TypeTableau;

void *creerTableau(void *arg)
{
    TypeTableau *a=(TypeTableau *)arg;
    a->tab=malloc(a->nbElements*sizeof(int));

    for(int i=0; i<a->nbElements; i++)
        a->tab[i]=rand()%100;
    return NULL;
}

void *chercheX(void *arg)
{
    TypeTableau *a=(TypeTableau *)arg;
    for(int i=0; i<a->nbElements; i++)
    {
        if(a->tab[i]==a->x)
            return (void *)1;
    }
    return (void *)0;
}

int main()
{
    pthread_t th1, th2;
    TypeTableau A;
    int resultat;
    srand(time(NULL));
    A.nbElements=rand()%10+1;

    pthread_create(&th1, NULL, creerTableau, &A);
    printf("Nombre à chercher: ");
    scanf("%d", &A.x);
    pthread_join(th1, NULL);

    for(int i=0; i<A.nbElements; i++)
        printf("%d  ", A.tab[i]);
    printf("\n");
    pthread_create(&th2, NULL, chercheX, &A);
    pthread_join(th2, (void **)&resultat);
    free(A.tab);
    return 0;
}